internal/server/storage/postgre.go

```diff
@@ -0,0 +1,329 @@
+	if err := postgre.Migrate(ctx); err != nil {
+		return nil, err
+	}
+
+	return &postgre, nil
+}
+
+func (s *PostgreStorage) Migrate(ctx context.Context) error {
+	var pgerr *pgconn.PgError
+
+	createType := `CREATE TABLE IF NOT EXISTS metric_types (
+        type VARCHAR(10) PRIMARY KEY
+	);`
+	_, err := s.db.ExecContext(ctx, createType)
+	if err != nil {
+		if errors.As(err, &pgerr) && pgerrcode.IsConnectionException(pgerr.Code) {
+			err = s.retryExecContext(ctx, createType)
+		}
+		if err != nil {
+			return err
+		}
+	}
+
+	insertType := `INSERT INTO metric_types (type) VALUES 
+		('counter'),
+		('gauge')
+		ON CONFLICT DO NOTHING
+		;`
+	_, err = s.db.ExecContext(ctx, insertType)
+	if err != nil {
+		if errors.As(err, &pgerr) && pgerrcode.IsConnectionException(pgerr.Code) {
+			err = s.retryExecContext(ctx, insertType)
+		}
+		if err != nil {
+			return err
+		}
+	}
+
+	createMetrics := `CREATE TABLE IF NOT EXISTS metrics (
+    	id serial PRIMARY KEY, 
+        name VARCHAR(255) UNIQUE NOT NULL,
+		mtype VARCHAR(10) references metric_types(type) NOT NULL,
+        delta BIGINT CHECK ((delta IS NOT NULL AND mtype = 'counter') OR (delta IS NULL AND mtype = 'gauge')),
+        value DOUBLE PRECISION CHECK ((value IS NOT NULL AND mtype = 'gauge') OR (value IS NULL AND mtype = 'counter'))
+	);`
+	_, err = s.db.ExecContext(ctx, createMetrics)
+	if err != nil {
+		if errors.As(err, &pgerr) && pgerrcode.IsConnectionException(pgerr.Code) {
+			err = s.retryExecContext(ctx, createMetrics)
+		}
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (s *PostgreStorage) All(ctx context.Context) ([]model.Metrics, error) {
+	var pgerr *pgconn.PgError
+
+	query := "SELECT name, mtype, delta, value FROM metrics;"
+	rows, err := s.db.QueryContext(ctx, query)
+	if err != nil {
+		if errors.As(err, &pgerr) && pgerrcode.IsConnectionException(pgerr.Code) {
+			rows, err = s.retryQueryContext(ctx, query)
+		}
+		if err != nil {
+			return nil, err
+		}
+	}
+	defer rows.Close()
+
+	var metrics []model.Metrics
+
+	for rows.Next() {
+		var (
+			name  string
+			mtype string
+			delta sql.NullInt64
+			value sql.NullFloat64
+		)
+
+		if err := rows.Scan(&name, &mtype, &delta, &value); err != nil {
+			return metrics, err
+		}
+
+		var m model.Metrics
+		if delta.Valid {
+			m = model.Metrics{
+				Name:  name,
+				MType: mtype,
+				Delta: &delta.Int64,
+			}
+		} else if value.Valid {
+			m = model.Metrics{
+				Name:  name,
+				MType: mtype,
+				Value: &value.Float64,
+			}
+		} else {
+			return metrics, ErrInvalidMetric
+		}
+		metrics = append(metrics, m)
+	}
+
+	if err = rows.Err(); err != nil {
+		return metrics, err
+	}
+	return metrics, nil
+}
+
+func (s *PostgreStorage) Get(ctx context.Context, name string) (model.Metrics, error) {
+	var pgerr *pgconn.PgError
+	var m model.Metrics
+
+	query := "SELECT name, mtype, delta, value FROM metrics WHERE name = $1;"
+
+	row := s.db.QueryRowContext(ctx, query, name)

```

Comment:
```suggestion
	row := s.db.QueryRowContext(ctx, query, name).Scan()
```
Тот же эффект, но меньше писанины. И, возможно, поход в бд лишний уйдет, если row.Eror не забирает данные из бд